### PR TITLE
Relaxed handle of corrupt png files

### PR DESCRIPTION
**Pull request by svenclaesson**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

WIP support for handing truncated png files

Implementing an option for PngDecoder, IgnoreCrcErrors
Tried it a bit but got stuck by the fact that IDAT and IEND is considered critical chunks and found no way of ignoring that fact if the file is truncated, IDAT is cut off and IEND is missing.

This file is truncated and i'm trying to get a partial result converted as libpng (chrome) and WIC can do.
280752616-57c5e277-a472-4f13-98f0-a2a238c0fe28

Maybe you have some pointers for me

See https://github.com/SixLabors/ImageSharp/issues/2570

### CODE REVIEW

**Comment by JimBobSquarePants** [comment]
@svenclaesson could you please explicitly invite me to contribute to your fork. I need to add the test image from above and due to a quirk in Git LFS permissions with GitHub I cannot push any images to your fork without being invited.

**Comment by JimBobSquarePants** [comment]
Thanks for the invite. @svenclaesson 

I've now reworked the decoder to allow granular control of CRC checking and also allow us to decode the provided image. 

Decode_InvalidDataChunkCrc_IgnoreCrcErrors_Rgba32_Issue_2589

File: tests/ImageSharp.Tests/Formats/ImageFormatManagerTests.cs [hunk 1]
```diff
@@ -43,7 +43,7 @@ public void IfAutoLoadWellKnownFormatsIsTrueAllFormatsAreLoaded()
         Assert.Equal(1, this.DefaultFormatsManager.ImageDecoders.Select(item => item.Value).OfType<PngDecoder>().Count());
         Assert.Equal(1, this.DefaultFormatsManager.ImageDecoders.Select(item => item.Value).OfType<BmpDecoder>().Count());
         Assert.Equal(1, this.DefaultFormatsManager.ImageDecoders.Select(item => item.Value).OfType<JpegDecoder>().Count());
-        Assert.Equal(1, this.DefaultFormatsManager.ImageDecoders.Select(item => item.Value).OfType<BmpDecoder>().Count());
+        Assert.Equal(1, this.DefaultFormatsManager.ImageDecoders.Select(item => item.Value).OfType<GifDecoder>().Count());
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/Formats/ImageFormatManagerTests.cs:46]: Really good spot!!

**Comment by svenclaesson** [comment]
Thanks for working on this

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs [hunk 2]
```diff
@@ -1771,21 +1810,27 @@ private bool TryReadChunk(Span<byte> buffer, out PngChunk chunk)
...
 
+        if (!this.TryReadChunkLength(buffer, out int length))
+        {
             // IEND
+            chunk = default;
             return false;
         }
 
-        while (length < 0 || length > (this.currentStream.Length - this.currentStream.Position))
+        while (length < 0)
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:1827]: @antonfirsov I did find a way to improve this!

**Pull request merged**